### PR TITLE
SNOW-1821505 Introduce disableOCSPChecks config option

### PR DESCRIPTION
**Pull request by sfc-gh-pfus**
### Description

SNOW-1821505 Introduced disableOCSPChecks and deprecated insecureMode.

### Checklist
- [x] Created tests which fail without the change (if possible)
- [x] Extended the README / documentation, if necessary

### CODE REVIEW

File: doc.go [hunk 1]
```diff
@@ -106,7 +106,7 @@ The following connection parameters are supported:
 
   - application: Identifies your application to Snowflake Support.
 
-  - insecureMode: false by default. Set to true to bypass the Online
+  - disableOCSPChecks: false by default. Set to true to bypass the Online
```
Comment (sfc-gh-dprzybysz) [inline doc.go:109]: we should keep both in docs, just `insecureMode` should be marked as deprecated

File: driver_test.go [hunk 2]
```diff
@@ -1793,7 +1793,7 @@ func TestOpenWithTransport(t *testing.T) {
 
 	// Test that transport override also works in insecure mode
 	countingTransport.requests = 0
-	config.InsecureMode = true
+	config.DisableOCSPChecks = true
```
Comment (sfc-gh-dprzybysz) [inline driver_test.go:1796]: I think we should have at least two tests (also for InsecureMode)

File: driver_test.go [hunk 3]
```diff
@@ -1793,7 +1793,7 @@ func TestOpenWithTransport(t *testing.T) {
 
 	// Test that transport override also works in insecure mode
```
Comment (sfc-gh-dprzybysz) [inline driver_test.go:1794]: should we change the comment also?

File: dsn.go [hunk 4]
```diff
@@ -79,8 +79,10 @@ type Config struct {
 	ExternalBrowserTimeout time.Duration // Timeout for external browser login
 	MaxRetryCount          int           // Specifies how many times non-periodic HTTP request can be retried
 
-	Application  string           // application name.
-	InsecureMode bool             // driver doesn't check certificate revocation status
+	Application       string // application name.
+	DisableOCSPChecks bool   // driver doesn't check certificate revocation status
+	// Deprecated: InsecureMode use DisableOCSPChecks instead
+	InsecureMode bool
```
Comment (sfc-gh-dprzybysz) [inline dsn.go:85]: let's keep the previous comment also

File: doc.go (hunk 1 above)
Comment (sfc-gh-pfus) [reply doc.go:109]: Ok, done.

File: driver_test.go (hunk 3 above)
Comment (sfc-gh-pfus) [reply driver_test.go:1810]: Ok, done.

File: driver_test.go (hunk 2 above)
Comment (sfc-gh-pfus) [reply driver_test.go:1812]: Done.

File: dsn.go (hunk 4 above)
Comment (sfc-gh-pfus) [reply dsn.go]: Done.

File: util.go [hunk 5]
```diff
@@ -339,3 +339,12 @@ func withLowerKeys[T any](in map[string]T) map[string]T {
 	}
 	return out
 }
+
+func getFirstIndexOfWithPrefix(in []string, prefix string) int {
```
Comment (sfc-gh-dheyman-1) [inline util.go:343]: `findByPrefix` ?

File: dsn.go [hunk 6]
```diff
@@ -637,7 +642,14 @@ func parseParams(cfg *Config, posQuestion int, dsn string) (err error) {
 // parseDSNParams parses the DSN "query string". Values must be url.QueryEscape'ed
 func parseDSNParams(cfg *Config, params string) (err error) {
 	logger.Infof("Query String: %v\n", params)
-	for _, v := range strings.Split(params, "&") {
+	paramsSlice := strings.Split(params, "&")
+	insecureModeIdx := getFirstIndexOfWithPrefix(paramsSlice, "insecureMode")
+	disableOCSPChecksIdx := getFirstIndexOfWithPrefix(paramsSlice, "disableOCSPChecks")
+	if insecureModeIdx > -1 && disableOCSPChecksIdx > -1 {
+		logger.Warn("duplicated insecureMode and disableOCSPChecks. disableOCSPChecks takes precedence")
+		paramsSlice = append(paramsSlice[:insecureModeIdx-1], paramsSlice[insecureModeIdx+1:]...)
```
Comment (sfc-gh-dheyman-1) [inline dsn.go:650]: is ellipisis operator needed here? Can't we just do `append(paramsSlice[:insecureModeIdx-1], paramsSlice[insecureModeIdx+1:])`?

**Review by sfc-gh-dheyman-1 — APPROVED** [review]

File: util.go (hunk 5 above)
Comment (sfc-gh-pfus) [reply util.go]: Yes, it sounds better.

File: dsn.go (hunk 6 above)
Comment (sfc-gh-pfus) [reply dsn.go:650]: No, because the function `append` uses vararg here.

File: driver_test.go (hunk 3 above)
Comment (sfc-gh-dprzybysz) [reply driver_test.go:1810]: it't not changed

Comment (sfc-gh-pfus) [reply driver_test.go:1810]: Actually it is - as you suggested, I left the `insecure mode` part, but added also `OCSP checks disabled`.

Comment (sfc-gh-dprzybysz) [reply driver_test.go]: Finally :)

**Pull request merged**